src/python/pants/init/logging.py

```diff
@@ -7,165 +7,162 @@
         )
 
-    def flush(self):
+    def flush(self) -> None:
         self.native.flush_log()
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return (
             f"NativeHandler(id={id(self)}, level={self.level}, filename={self.native_filename}, "
             f"stream={self.stream})"
         )
 
 
-def create_native_pantsd_file_log_handler(level, native, native_filename):
-    fd = native.setup_pantsd_logger(native_filename, get_numeric_level(level))
+def create_native_pantsd_file_log_handler(
+    log_level: LogLevel, native: Native, native_filename: str
+) -> NativeHandler:
+    fd = native.setup_pantsd_logger(native_filename, log_level.level)
     ExceptionSink.reset_interactive_output_stream(os.fdopen(os.dup(fd), "a"))
-    return NativeHandler(level, native, native_filename=native_filename)
-
-
-def create_native_stderr_log_handler(level, native, stream=None):
-    try:
-        native.setup_stderr_logger(get_numeric_level(level))
-    except Exception as e:
-        print(f"Error setting up pantsd logger: {e!r}", file=sys.stderr)
-        raise e
-
-    return NativeHandler(level, native, stream)
+    return NativeHandler(log_level, native, native_filename=native_filename)
 
 
-# TODO This function relies on logging._checkLevel, which is private.
-# There is currently no good way to convert string levels to numeric values,
-# but if there is ever one, it may be worth changing this.
-def get_numeric_level(level):
-    return logging._checkLevel(level)
-
+@overload
+def setup_logging(
+    log_level: LogLevel,
+    *,
+    native: Native,
+    log_dir: str,
+    console_stream: Optional[TextIO] = None,
+    scope: Optional[str] = None,
+    log_name: Optional[str] = None,
+    warnings_filter_regexes: Optional[List[str]] = None,
+) -> FileLoggingSetupResult:
+    ...
 
-@contextmanager
-def encapsulated_global_logger():
-    """Record all the handlers of the current global logger, yield, and reset that logger.
 
-    This is useful in the case where we want to easily restore state after calling setup_logging.
-    For instance, when DaemonPantsRunner creates an instance of LocalPantsRunner it sets up specific
-    nailgunned logging, which we want to undo once the LocalPantsRunner has finished runnnig.
-    """
-    global_logger = logging.getLogger()
-    old_handlers = copy.copy(global_logger.handlers)
-    try:
-        yield
-    finally:
-        new_handlers = global_logger.handlers
-        for handler in new_handlers:
-            global_logger.removeHandler(handler)
-        for handler in old_handlers:
-            global_logger.addHandler(handler)
+@overload
+def setup_logging(
+    log_level: LogLevel,
+    *,
+    native: Native,
+    log_dir: None = None,
+    console_stream: Optional[TextIO] = None,
+    scope: Optional[str] = None,
+    log_name: Optional[str] = None,
+    warnings_filter_regexes: Optional[List[str]] = None,
+) -> None:
+    ...
```

Comment:
I don't quite understand the `...` syntax. I infer it means "do the same thing as the function I'm overloading, but with slightly different output types. If so, couldn't we return a `Union` or something else? If we can't, could you add a comment explaining why we need this here?

Reply:
Check out https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading. Function overloading is a new type hints feature where you're able to get more precise with type hints when you know you were given a certain input. The `...` means "use the normal implementation".

The definitions of overloaded methods are very ugly because there's so much boilerplate, but it makes for fantastic type inference downstream.

Reply:
> If so, couldn't we return a Union or something else?

We do. `Optional[X]` desugars to `Union[None, X]`. Overloading allows tracking which output type you get for which input types. That's a trick you can't do in many type systems since overloads generally don't allow variance in the return type, just the parameter lists.

> If we can't, could you add a comment explaining why we need this here?

Eric pointed out the homework. In this case, since mypy is widespread in the codebase, I'd consider this just part of the learning curve in working in the codebase. I.E.: you must know python, you must know type-hints / mypy. It is true that `@overload` is somewhat esoteric:
```
$ git grep "@overload" | wc -l
10
```

But it is well documented.

Reply:
TIL, thanks! I'm still not really happy with overloading as a concept, but this is definitely not the time or the place :) If this is the way to do it, this is the way to do it :) Thanks both for the pointers!

Reply:
There is another juicy, terrifying spot that may need to be typed with `@overloads` nearby:
https://github.com/pantsbuild/pants/blob/1580ab213f13314f6977768cc2d64f4d2ceb19cd/src/python/pants/pantsd/pants_daemon.py#L167-L194

That difference in resulting `PantsDaemon` instances though is so radical that its probably much better to split out two types a PartialPantsDaemon and a fully function PantsDaemon. Right now its a minefield of members expected to be non-`None` that is only ensured by carefully calling just the right methods only on partial / full `PantsDaemon` instances.